File: 3633.earliest-finish-time-for-land-and-water-rides-i/solution.kt
class Solution {
    fun earliestFinishTime(landStartTime: IntArray, landDuration: IntArray, waterStartTime: IntArray, waterDuration: IntArray): Int {
        return minOf(
            optimize(landStartTime, landDuration, waterStartTime, waterDuration),
            optimize(waterStartTime, waterDuration, landStartTime, landDuration)
        )
    }

    fun optimize(s1: IntArray, d1: IntArray, s2: IntArray, d2: IntArray): Int {
        val t1 = (0 until s1.size).minOf { s1[it] + d1[it] }
        return (0 until s2.size).minOf { maxOf(t1, s2[it]) + d2[it] }
    }
}

File: 3633.earliest-finish-time-for-land-and-water-rides-i/solution.go
func earliestFinishTime(landStartTime []int, landDuration []int, waterStartTime []int, waterDuration []int) int {
    t1 := optimize(landStartTime, landDuration, waterStartTime, waterDuration)
    t2 := optimize(waterStartTime, waterDuration, landStartTime, landDuration)
    return min(t1, t2)
}

func optimize(s1, d1, s2, d2 []int) int {
    t1, t2 := math.MaxInt, math.MaxInt

    for i := range s1 {
        t1 = min(s1[i] + d1[i], t1)
    }
    for i := range s2 {
        t2 = min(max(s2[i], t1) + d2[i], t2)
    }

    return t2
}